README.rst
=======
pyaavso
=======

.. image:: https://requires.io/github/zsiciarz/pyaavso/requirements.svg?branch=master
    :target: https://requires.io/github/zsiciarz/pyaavso/requirements/?branch=master
    :alt: Requirements Status

.. image:: https://img.shields.io/pypi/v/pyaavso.svg
    :target: https://pypi.python.org/pypi/pyaavso/
    :alt: Latest PyPI version

.. image:: https://img.shields.io/pypi/dm/pyaavso.svg
    :target: https://pypi.python.org/pypi/pyaavso/
    :alt: Number of PyPI downloads

.. image:: https://img.shields.io/pypi/pyversions/pyaavso.svg
    :target: https://pypi.python.org/pypi/pyaavso/
    :alt: Supported Python versions

.. image:: https://img.shields.io/pypi/wheel/pyaavso.svg
    :target: https://pypi.python.org/pypi/pyaavso/
    :alt: Wheel Status

.. image:: https://travis-ci.org/zsiciarz/pyaavso.svg?branch=master
    :target: https://travis-ci.org/zsiciarz/pyaavso

.. image:: https://coveralls.io/repos/zsiciarz/pyaavso/badge.svg?branch=master
    :target: https://coveralls.io/r/zsiciarz/pyaavso?branch=master

**pyaavso** is a Python library for working with
`AAVSO <http://www.aavso.org>`_ (American Association of Variable Star
Observers) data. The library is compatible with both Python 2.7 and 3.3+.

Features
--------

* reading and writing variable star observations in AAVSO's
  `Visual File Format`_
* downloading all observation data for a given observer

.. _`Visual File Format`: http://www.aavso.org/aavso-visual-file-format

Installation
------------

Use ``pip`` to install latest release available at PyPI::

    pip install pyaavso

Usage
-----

The following code uses ``VisualFormatWriter`` to report a single
observation of **SS Cyg** between the outbursts.

    >>> from pyaavso.formats import VisualFormatWriter
    >>> observer_code = 'XYZ'
    >>> with open('data.txt', 'wb') as fp:
    ...     writer = VisualFormatWriter(fp, observer_code)
    ...     writer.writerow({
    ...         'name': 'SS CYG',
    ...         'date': '2450702.1234',
    ...         'magnitude': '<11.0',
    ...         'comp1': '110',
    ...         'chart': '070613',
    ...     })

The ``data.txt`` file can be now submitted to AAVSO.

Resources
---------

 * `Documentation <http://pyaavso.rtfd.org>`_
 * `Issue tracker <https://github.com/zsiciarz/pyaavso/issues>`_
 * `CI server <https://travis-ci.org/zsiciarz/pyaavso>`_

Author
------

 * `Zbigniew Siciarz <http://siciarz.net>`_ (zbigniew at siciarz dot net)

License
-------

pyaavso is free software, licensed under the MIT/X11 License. A copy of
the license is provided with the source code in the LICENSE file.


CHANGES.rst
=========
Changelog
=========

Unreleased
==========

* none yet

0.2.0
=====

* first **Python 3-only** release!
* Python 3.5 and 3.6 support

0.1.5
=====

* fixed VisualFormatReader bug on Python 3 when input is bytes, not string

0.1.4
=====

* minor packaging and documentation fixes

0.1.3
=====

* Python 3.4 compatibility
* more specific Python version classifiers in setup.py

0.1.2
=====

* added wheel distribution

0.1.1
=====

* less memory-hungry VisualFormatReader

0.1.0
=====

* initial release


pyaavso/formats/visual.py
import csv
import itertools

import pyaavso


class FormatException(Exception):
    """
    Raised when the data does not conform to AAVSO format specification.
    """


class VisualFormatWriter(object):
    """
    A class responsible for writing observation data in AAVSO
    `Visual File Format`_.

    The API here mimics the ``csv`` module in Python standard library.

    To write your observations into the data file, you first need to create
    the writer, passing to it the destination file and your observer code.
    Then call :py:meth:`~VisualFormatWriter.writerow` for every single
    observation, for example:

        >>> with open('data.txt', 'wb') as fp:
        ...     writer = VisualFormatWriter(fp, 'XYZ')
        ...     writer.writerow({
        ...         'name': 'SS CYG',
        ...         'date': '2450702.1234',
        ...         'magnitude': '<11.1',
        ...         'comment_code': '',
        ...         'comp1': '110',
        ...         'comp2': '113',
        ...         'chart': '070613',
        ...         'notes': 'This is a test',
        ...     })

    .. _`Visual File Format`: http://www.aavso.org/aavso-visual-file-format
    """

    def __init__(self, fp, observer_code, *, delimiter=',', date_format='JD',
                 obstype='Visual'):
        """
        Creates the writer which will write observations into the file-like
        object given in first parameter. The only other required parameter
        is the official AAVSO-assigned observer code.

        :param fp: file-like object to write observations into
        :param observer_code: AAVSO observer code
        :param delimiter: field delimiter (set as DELIM header)
        :param date_format: observation date format (one of *JD* or *Excel*)
        :param obstype: observation type (*Visual* or *PTG*)
        """
        self.observer_code = observer_code
        self.date_format = date_format
        self.obstype = obstype
        fp.write('#TYPE=Visual\n')
        fp.write('#OBSCODE=%s\n' % observer_code)
        fp.write("#SOFTWARE=pyaavso %s\n" % pyaavso.get_version())
        fp.write("#DELIM=%s\n" % delimiter)
        fp.write("#DATE=%s\n" % date_format.upper())
        fp.write("#OBSTYPE=%s\n" % obstype)
        self.writer = csv.writer(fp, delimiter=delimiter)

    def writerow(self, observation_data):
        """
        Writes a single observation to the output file.

        If the ``observation_data`` parameter is a dictionary, it is
        converted to a list to keep a consisted field order (as described
        in format specification). Otherwise it is assumed that the data
        is a raw record ready to be written to file.

        :param observation_data: a single observation as a dictionary or list
        """
        if isinstance(observation_data, (list, tuple)):
            row = observation_data
        else:
            row = self.dict_to_row(observation_data)
        self.writer.writerow(row)

    @classmethod
    def dict_to_row(cls, observation_data):
        """
        Takes a dictionary of observation data and converts it to a list
        of fields according to AAVSO visual format specification.

        :param cls: current class
        :param observation_data: a single observation as a dictionary
        """
        row = []
        row.append(observation_data['name'])
        row.append(observation_data['date'])
        row.append(observation_data['magnitude'])
        comment_code = observation_data.get('comment_code', 'na')
        if not comment_code:
            comment_code = 'na'
        row.append(comment_code)
        comp1 = observation_data.get('comp1', 'na')
        if not comp1:
            comp1 = 'na'
        row.append(comp1)
        comp2 = observation_data.get('comp2', 'na')
        if not comp2:
            comp2 = 'na'
        row.append(comp2)
        chart = observation_data.get('chart', 'na')
        if not chart:
            chart = 'na'
        row.append(chart)
        notes = observation_data.get('notes', 'na')
        if not notes:
            notes = 'na'
        row.append(notes)
        return row


class VisualFormatReader(object):
    """
    A class to read observations from file in AAVSO
    `Visual File Format`_.

    .. _`Visual File Format`: http://www.aavso.org/aavso-visual-file-format

    The reader API is also based on ``csv`` Python module. You create a reader
    instance by passing a file-like object in the constructor. This will
    read all the data and validate required headers. Then the reader object
    can be used to iterate over observation data.

    A short example:

        >>> with open('data.txt', 'rb') as fp:
        ...     reader = VisualFormatReader(fp)
        ...     for observation in reader:
        ...         print('%(name)s %(magnitude)s' % observation)
        SS Cyg 10.0
        RZ Cas 6.4
    """

    def __init__(self, fp):
        """
        Creates the reader instance and reads file headers.

        Raises :py:exc:`~pyaavso.format.visual.FormatException` when any of
        the required headers could not be found in input. The following header
        parameters are required:

         * *TYPE* - always 'Visual', yet must be specified in file
         * *OBSCODE* - official AAVSO-assigned observer code
         * *DATE* - date format, must be one of 'JD' or 'Excel'

        Other headers described in AAVSO specification have reasonable default
        values, eg. the default delimiter is a comma, when not specified
        in headers. Without the *OBSTYPE* header, observations are assumed
        to be visual.

        :param fp: a file-like object from which data will be read
        """
        headers = {}
        for line in fp:
            if isinstance(line, bytes):
                line = line.decode('utf-8')
            line = line.strip()
            if line and line[0] == '#' and '=' in line:
                header_str = line[1:]
                key, value = header_str.split('=', 1)
                headers[key] = value
            elif line and line[0] != '#':
                # first non-comment line marks the beggining of data section
                break
        # validate headers
        if 'TYPE' not in headers:
            raise FormatException('TYPE parameter is required')
        try:
            self.observer_code = headers['OBSCODE']
        except KeyError:
            raise FormatException('OBSCODE parameter is required')
        try:
            self.date_format = headers['DATE']
        except KeyError:
            raise FormatException('DATE parameter is required')
        self.software = headers.get('SOFTWARE', '')
        self.delimiter = str(headers.get('DELIM', ','))
        self.obstype = headers.get('OBSTYPE', 'Visual')
        # prepend the peeked line and continue iteration
        data = itertools.chain([line], fp)
        self.reader = csv.reader(data, delimiter=self.delimiter)

    def __iter__(self):
        for row in self.reader:
            yield self.row_to_dict(row)

    @classmethod
    def row_to_dict(cls, row):
        """
        Converts a raw input record to a dictionary of observation data.

        :param cls: current class
        :param row: a single observation as a list or tuple
        """
        comment_code = row[3]
        if comment_code.lower() == 'na':
            comment_code = ''
        comp1 = row[4]
        if comp1.lower() == 'na':
            comp1 = ''
        comp2 = row[5]
        if comp2.lower() == 'na':
            comp2 = ''
        chart = row[6]
        if chart.lower() == 'na':
            chart = ''
        notes = row[7]
        if notes.lower() == 'na':
            notes = ''
        return {
            'name': row[0],
            'date': row[1],
            'magnitude': row[2],
            'comment_code': comment_code,
            'comp1': comp1,
            'comp2': comp2,
            'chart': chart,
            'notes': notes,
        }
